components/channel/include/channel.h

```diff
@@ -0,0 +1,232 @@
+
+/**
+ * channel_init_consumer - helper function to initialize channel object as consumer
+ * @ch: pointer to channel object to initialize
+ * @identifier: identifier string of the channel
+ * @queue: queue handle to send input to 
+ */
+static
+void
+inline __attribute__((always_inline))
+channel_init_consumer
+(Channel_consumer *ch, const char *identifier, QueueHandle_t queue)
+{
+    return channel_init(ch, identifier, queue, queueSEND_TO_BACK, &xQueueGenericSend);
+}
+
+/**
+ * channel_init_producer - helper function to initialize channel object as producer
+ * @ch: pointer to channel object to initialize
+ * @identifier: identifier string of the channel
+ */
+static
+void
+inline __attribute__((always_inline))
+channel_init_producer
+(Channel_producer *ch, const char *identifier)
+{
+    return channel_init(ch, identifier, NULL, 0, NULL);
+}
+
+/**
+ * channel_setContext - helper function to change the context the callback acts on
+ * @ch: pointer to channel object to modify
+ * @ctx: pointer to new context
+ */
+static
+void
+inline __attribute__((always_inline))
+channel_setContext 
+(Channel *ch, void *ctx)
+{
+    ch->ctx = ctx;
+}
+
+/**
+ * channel_setCallback - helper function to change the callback
+ * @ch: pointer to channel object to modify
+ * @callback: function to register as callback
+ */
+static
+void 
+inline __attribute__((always_inline))
+channel_setCallback
+(Channel *ch, const Channel_callback callback)
+{
+    ch->callback = callback;
+}
+
+/**
+ * reset_channel - removes a channel input or output from the message structure
+ * @ch: channel element to reset
+ */
+void
+channel_reset
+(Channel *ch);
+
+/**
+ * channel_send - sends a message to a single channel object
+ * @ch: the channel structure that holds informations for this channel
+ * @data: the pointer to the message
+ * @timeout: how many time-slices the callback may block before aborting
+ * @return: returns pdPass (=1) if send was successful, if non-one value is returned, callback failed
+ * 
+ * Calls the callback function on the channel.
+ */
+BaseType_t
+channel_send
+(const Channel_consumer *ch, const void *data, const TickType_t timeout);
+
+/**
+ * channel_broadcast_init - helper function to initialize broadcast handler
+ * @handle: pointer to broadcast object
+ * @ch: pointer to channel object initiating the broadcast
+ * @data: pointer to data to be transferred
+ * @timeout: number of time-slices each receiver may block 
+ */
+static
+void
+inline __attribute__((always_inline))
+channel_broadcast_init
+(Channel_broadcast *handle, const Channel_producer *ch, const void *data, const TickType_t timeout)
+{
+    handle->ch = (void*) ch;
+    handle->pos = (void*) ch;
+    handle->data = (void*) data;
+    handle->timeout = timeout;
+    handle->elapsed = 0;
+}
+
+/**
+ * channel_broadcast_timeout - helper function to determine if broadcast timed out
+ * @handle: pointer to broadcast object
+ */
+static
+bool
+inline __attribute__((always_inline))
```

Comment:
```suggestion
__attribute__((always_inline))
```

Reply:
> GCC does not inline any functions when not optimizing unless you specify the ‘always_inline’ > attribute for the function, like this:
>
>/* Prototype.  */
> `inline void foo (const char) __attribute__((always_inline));`

see https://gcc.gnu.org/onlinedocs/gcc/Inline.html

Reply:
> > ** 1) As I understood the description of always_inline, the inline
> keyword in the above example is obsolete, right?
> [According to my tests this seems to be the case, but want to be sure :-)]

> If you use the always_inline attribute, then, yes, you do not need to
> also use the inline keyword.  The inline keyword is not in general
> obsolete.

see: https://gcc.gnu.org/legacy-ml/gcc-help/2007-01/msg00051.html

Reply:
You are right, it seems unnecessary, but will throw warnings:
```
components/channel/include/channel.h:97:1: warning: always_inline function might not be inlinable [-Wattributes]
 channel_init_consumer
 ^~~~~~~~~~~~~~~~~~~~~
esp32-core/components/channel/include/channel.h:97:1: warning: 'channel_init_consumer' defined but not used [-Wunused-function]
```
we could remove the attribute, since we always build with optimizations and then no combination yields a symbol